core/src/main/java/org/elasticsearch/common/util/concurrent/QueueResizingEsThreadPoolExecutor.java

```diff
@@ -0,0 +1,161 @@
+        this.tasksPerWindow = tasksPerWindow;
+        this.startNs = System.nanoTime();
+        this.minQueueSize = minQueueSize;
+        this.maxQueueSize = maxQueueSize;
+    }
+
+    @Override
+    protected void doExecute(final Runnable command) {
+        // we are submitting a task, it has not yet started running (because super.excute() has not
+        // been called), but it could be immediately run, or run at a later time. We need the time
+        // this task entered the queue, which we get by creating a TimedRunnable, which starts the
+        // clock as soon as it is created.
+        super.doExecute(this.runnableWrapper.apply(command));
+    }
+
+    static int calculateL(final long totalRuntime, final int taskCount,
+                          final long totalTaskNanos, final String name,
+                          final Logger logger) {
+        assert totalRuntime > 0 : "cannot calculate for instantaneous tasks";
+        assert taskCount > 0 : "cannot calculate for no tasks";
+        // There is no set execution time, instead we adjust the time window based on the
+        // number of completed tasks, so there is no backwards thread required to update the
+        // queue size at a regular interval. This means we need to calculate our λ by the
+        // total runtime, rather than a fixed interval.
+
+        // λ = tasks per time period - total tasks divided by measurement time
+        final double lambda = (double) taskCount / totalRuntime;
+
+        // W = avg task execution time - total task execution divided by total task count
+        final double w = totalTaskNanos / taskCount;
+
+        // L = λ * W
+        final int l = (int) (lambda * w);
+        logger.debug("estimated optimal queue size for [{}] threadpool: " +
+                        "[{}](L) = {}(λ) * {}(W)", name, l, lambda, w);
+        return l;
+    }
+
+    @Override
+    protected void afterExecute(Runnable r, Throwable t) {
+        super.afterExecute(r, t);
+        // A task has been completed, it has left the building. We should now be able to get the
+        // total time as a combination of the time in the queue and time spent running the task. We
+        // only want runnables that did not throw errors though, because they could be fast-failures
+        // that throw off our timings, so only check when `t` is null.
+        if (t == null) {
+            assert r instanceof TimedRunnable : "expected only TimedRunnables in queue";
+            long taskNanos = ((TimedRunnable) r).getTotalNanos();
+            long totalNanos = totalTaskNanos.addAndGet(taskNanos);
+            if (taskCount.incrementAndGet() == this.tasksPerWindow) {
+                final long totalRuntime = System.nanoTime() - this.startNs;
+                // Reset the start time for all tasks
+                startNs = System.nanoTime();
+                // Reset the task count and time back to 0 to start collecting task times again
+                taskCount.set(0);
```

Comment:
if you set this to `0` here we could get concurrent executions of this method since there might be enough tasks being executed and finished concurrently. I think you should set it to `0` once this block has finished. maybe even in a finally block that would highlight it place where we do it and gives you a good place for a comment